controller-server/src/main/java/com/yahoo/vespa/hosted/controller/persistence/CuratorDb.java

```diff
@@ -331,33 +333,42 @@ public void removeTenant(TenantName name) {
 
     // -------------- Application ---------------------------------------------
 
-    public void writeApplication(Instance instance) {
+    public void writeInstance(Instance instance) {
         curator.set(applicationPath(instance.id()), asJson(instanceSerializer.toSlime(instance)));
+        curator.set(instancePath(instance.id()), asJson(instanceSerializer.toSlime(instance)));
     }
 
-    public Optional<Instance> readApplication(ApplicationId application) {
-        return readSlime(applicationPath(application)).map(instanceSerializer::fromSlime);
+    public Optional<Instance> readInstance(ApplicationId application) {
+        return readSlime(instancePath(application)).or(() -> readSlime(applicationPath(application)))
+                                                   .map(instanceSerializer::fromSlime);
     }
 
-    public List<Instance> readApplications() {
-        return readApplications(ignored -> true);
+    public List<Instance> readInstances() {
+        return readInstances(ignored -> true);
     }
 
-    public List<Instance> readApplications(TenantName name) {
-        return readApplications(application -> application.tenant().equals(name));
+    public List<Instance> readInstances(TenantName name) {
+        return readInstances(application -> application.tenant().equals(name));
     }
 
-    private List<Instance> readApplications(Predicate<ApplicationId> applicationFilter) {
-        return curator.getChildren(applicationRoot).stream()
-                      .map(ApplicationId::fromSerializedForm)
-                      .filter(applicationFilter)
-                      .map(this::readApplication)
-                      .flatMap(Optional::stream)
-                      .collect(collectingAndThen(Collectors.toList(), Collections::unmodifiableList));
+    private Stream<ApplicationId> readInstanceIds() {
+        return Stream.concat(curator.getChildren(applicationRoot).stream()
+                                    .filter(id -> id.split(":").length == 3),
+                             curator.getChildren(instanceRoot).stream())
+                     .distinct()
+                     .map(ApplicationId::fromSerializedForm);
+    }
+
+    private List<Instance> readInstances(Predicate<ApplicationId> instanceFilter) {
+        return readInstanceIds().filter(instanceFilter)
+                                .map(this::readInstance)
+                                .flatMap(Optional::stream)
+                                .collect(Collectors.toUnmodifiableList());
     }
 
-    public void removeApplication(ApplicationId application) {
-        curator.delete(applicationPath(application));
+    public void removeInstance(ApplicationId id) {
+        curator.delete(applicationPath(id));
```

Comment:
This is a bug waiting to happen ... perhaps put a big fat warning comment here?

Reply:
Can do that.